pkg/dataobj/querier/store.go

```diff
@@ -632,3 +723,277 @@ func parseShards(shards []string) (logql.Shard, error) {
+	return predicate, expr
+}
+
+func buildLogMessagePredicateFromPipeline(expr syntax.LogSelectorExpr) (dataobj.LogsPredicate, syntax.LogSelectorExpr) {
+	// Check if expr is a PipelineExpr, other implementations have no stages
+	pipelineExpr, ok := expr.(*syntax.PipelineExpr)
+	if !ok {
+		return nil, expr
+	}
+
+	var (
+		predicate       dataobj.LogsPredicate
+		remainingStages = make([]syntax.StageExpr, 0, len(pipelineExpr.MultiStages))
+		appendPredicate = func(p dataobj.LogsPredicate) {
+			if predicate == nil {
+				predicate = p
+			} else {
+				predicate = dataobj.AndPredicate[dataobj.LogsPredicate]{
+					Left:  predicate,
+					Right: p,
+				}
+			}
+		}
+	)
+
+Outer:
+	for i, stage := range pipelineExpr.MultiStages {
+		switch s := stage.(type) {
+		case *syntax.LineFmtExpr:
+			// modifies the log line, break early as we cannot apply any more predicates
+			remainingStages = append(remainingStages, pipelineExpr.MultiStages[i:]...)
+			break Outer
+
+		case *syntax.LineFilterExpr:
+			// Convert the line filter to a predicate
+			f, err := s.Filter()
+			if err != nil {
+				remainingStages = append(remainingStages, s)
+				continue
+			}
+
+			// Create a line filter predicate
+			appendPredicate(dataobj.LogMessageFilterPredicate{
+				Keep: func(line []byte) bool {
+					return f.Filter(line)
+				},
+			})
+
+		default:
+			remainingStages = append(remainingStages, s)
+		}
+	}
+
+	if len(remainingStages) == 0 {
+		return predicate, pipelineExpr.Left // return MatchersExpr
+	}
+	pipelineExpr.MultiStages = remainingStages
+
+	return predicate, pipelineExpr
+}
+
+func buildMetadataFilterPredicateFromSampleExpr(expr syntax.SampleExpr, metadataColumns map[string]struct{}) (dataobj.LogsPredicate, syntax.SampleExpr) {
+	var (
+		predicate dataobj.LogsPredicate
+		skip      bool
+	)
+	expr.Walk(func(e syntax.Expr) {
+		switch e := e.(type) {
+		case *syntax.BinOpExpr:
+			// we might not encounter BinOpExpr at this point since the lhs and rhs are evaluated separately?
+			skip = true
+			return
+		case *syntax.RangeAggregationExpr:
+			if skip {
+				return
+			}
+
+			predicate, e.Left.Left = buildMetadataFilterPredicateFromPipeline(e.Left.Left, metadataColumns)
+		}
+	})
+
+	return predicate, expr
+}
+
+func buildMetadataFilterPredicateFromPipeline(expr syntax.LogSelectorExpr, metadataColumns map[string]struct{}) (dataobj.LogsPredicate, syntax.LogSelectorExpr) {
+	pipelineExpr, ok := expr.(*syntax.PipelineExpr)
+	if !ok {
+		return nil, expr
+	}
+
+	var predicate dataobj.LogsPredicate
+	appendPredicate := func(p dataobj.LogsPredicate) {
+		if predicate == nil {
+			predicate = p
+		} else {
+			predicate = dataobj.AndPredicate[dataobj.LogsPredicate]{Left: predicate, Right: p}
+		}
+	}
+
+	remainingStages := make([]syntax.StageExpr, 0, len(pipelineExpr.MultiStages))
+
+Outer:
+	for i, stage := range pipelineExpr.MultiStages {
+		switch stage := stage.(type) {
+		case *syntax.LabelFmtExpr, *syntax.LineParserExpr, *syntax.LogfmtParserExpr,
```

Comment:
Maybe pattern here too? That can add to the label set

Reply:
https://github.com/grafana/loki/blob/21bf5ee5a77c6dba751d8a8189215ee348ebf0a3/pkg/logql/syntax/syntax.y.go#L1371-L1390
LineParserExpr should cover it